Refactor address state validation

cedum opened the pull request:
**Description**

Deprecates and refactors state validation by extracting out the normalization and validation logic into a separate service-like object, which might be also user-provided, for example:
```
Spree::Address.state_validator_class = UserProvided::Address::StateValidator
```

The following private methods have been deprecated from `Spree::Address` model:
- `state_validate`
- `validate_state_matches_country`

This PR introduces the following configuration flag `Spree::Config.use_legacy_address_state_validator` that enables to switch from the old state validation behavior to the new in a smooth way.
For newly generated applications, this flag is set to `false`, which means it will use the new state validator class.
For existing applications upgrading to Solidus v2.11 it is set to `true`, hence keeping the same behavior as in the previous Solidus versions.

**Checklist:**
- [ ] Write a recap in this PR description with the normalization and validation rules
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)

cedum commented:
Moved the PR back to WIP 🚧. It requires some more work. I'll try to make it less break. Besides this, by digging deeper into it I see I misinterpreted some state behaviours and also introduced a few bugs 😅

kennyadsl commented on core/app/models/spree/address.rb:
```diff
@@ -8,15 +8,22 @@ module Spree
-    validate :state_validate
-    validate :validate_state_matches_country
+    validate do
+      self.class.state_normalizer_class.new(self).perform!
```
Normalization is also happening into the validation, right? Why it is needed to perform it here as well?

cedum replied:
The original idea was to move it here from the validator service object so they can be completely decoupled. Just forgot to remove it from there (see also the comment bellow related to this one)

kennyadsl commented on core/app/models/spree/address/state_validator.rb:
```diff
@@ -0,0 +1,50 @@
+
+      # normalize state and state_name attributes before
+      # performing the validation
+      Spree::Address.state_normalizer_class.new(address).perform!
```
What about passing this class to the `initialize` of this class?

```ruby
def initialize(address, address_normalizer_class = Spree::Address.state_normalizer_class)

# and in the caller
self.class.state_validator_class.new(self, self.class.state_normalizer_class).perform!
```

This way it will still be possible to change the class as an address level preference but we can also have more flexibility if needed for some customization or testing needs.

cedum replied:
This wasn't intended to be here. I have moved it out into `validate` block in `Spree::Address`. Forgot to remove it from here :|

kennyadsl commented on core/app/models/spree/address/state_normalizer.rb:
```diff
@@ -0,0 +1,45 @@
+      @address = address
+    end
+
+    def perform!
```
Personal preference, but I'd use `def call` here. Did you have a specific reason to use `perform!` instead?

cedum replied:
I have followed the convention of other service-like objects in Solidus that are using `perform` as their main action method. I have no preferences though.
Probably it's worth defining a project-wise convention for this kind of objects

cedum commented:
Ready for a second review 🙏 

@kennyadsl I've researched a few methods to deprecate them by showing a warning message only if they have been monkey patched, but the final solution was quite hacky and not working in all cases since there're multiple ways to patch them.
Being both of them private methods, you think it's worth the effort to deprecate them before removing? I've invested some time into it but I wasn't able to find a +/- clean way to deprecate them.

cedum commented:
Rebased against master and resolved conflicts

aldesantis commented:
@cedum can you provide some info about the current state of the deprecation warning in this PR? It looks like everything should work properly if the private methods are being overridden, am I mistaken?

aldesantis commented on core/app/models/spree/address.rb:
```diff
@@ -8,15 +8,31 @@ module Spree
+
+      %w[state_validate validate_state_matches_country].each do |method|
+        if self.class.method_defined?(method)
+          Rails.logger.warn(
```
What if we called the custom methods instead of the new behavior in case the methods have been added by the user in a decorator? We may even do that in the `state_normalizer_class` and `state_validator_class` themselves.

kennyadsl commented:
I recall we decided to merge this in 3.0 but I think it wasn't a good decision since, if possible, we should make this optional and deprecate the old behavior in 2.11. This way people have time to fix this in their application before the 3.0 upgrade.

kennyadsl commented on core/app/models/spree/address.rb:
```diff
@@ -218,6 +228,12 @@ def validate_name
     end
 
     def state_validate
+      Spree::Deprecation.warn \
```
Until the preference is not changed users will see 3 deprecation warnings. It would be awesome to find a way to only print one deprecation warning in that case, but keep the individual ones if these methods are called from somewhere else. I'm not sure it's a good idea but what about passing an arg to these deprecated methods to conditionally silence the deprecation warning?

This way we can do something like:

```
      if Spree::Config.legacy_address_state_validator
        state_validate(silence_deprecation: true)
        validate_state_matches_country(silence_deprecation: true)
      else
```

Users still see the main deprecation warning if using default code and the specific ones if they used `state_validate` or `validate_state_matches_country` in other ways. I'm not particularly concerned about changing the arity of those methods since they'll go away soon. 

What's your thoughts?

cedum replied:
👍 agree!
By introducing a new parameter it will break any existing prepended patches of these methods. I'm not aware of a way to introduce a new parameter w/o being breaking. I'm looking if there's another way

kennyadsl replied:
Maybe just setting an accessor on the class?

cedum replied:
Another way might be to wrap the two calls in a `silence` block, this avoids polluting the Address class API with new attributes/methods. The downside is it will silence any eventual deprecations generated by these two calls.
What do you think?

```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    Spree::Deprecation.silence do
      state_validate
      validate_state_matches_country
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```

kennyadsl replied:
It's more elegant for sure but I think (might be worth a double-check) we are not using `Spree::Deprecation.silence` anywhere in the codebase (outside specs). I'm not completely sure why to be honest and if this could cause issues that we can't anticipate at the moment. Anyway, for now this seems like the best solution let's go head with this and let's see what the other core team member think about it.

cedum replied:
just discovered that `silence` seems to not be thread-safe 🤦 Here's a simple test case that sometimes leaves silenced class attribute set to `true`: https://gist.github.com/cedum/8315ebdd59781ee7fcc7f452a405542f
I'm refactoring this to a simple instance variable switch 😄, something like this:
```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    begin
      @silence_state_deprecations = true
      state_validate
      validate_state_matches_country
    ensure
      @silence_state_deprecations = false
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```

kennyadsl commented on core/lib/spree/testing_support/dummy_app.rb:
```diff
@@ -119,6 +119,7 @@ class Application < ::Rails::Application
   config.raise_with_invalid_currency = false
   config.redirect_back_on_unauthorized = true
   config.run_order_validations_on_order_updater = true
+  config.legacy_address_state_validator = false
```
Can we alphabetize this new preference?

kennyadsl reviewed (approved):
👍

kennyadsl commented:
@cedum we discussed this in the core meeting and we agreed that it would be much better if we have the fix for #2417 and #3098 in a different commit, just for documentation in case someone cannot update and wants to just understand what changed to backport that in their own application. Do you think it's doable?

cedum commented:
> @cedum we discussed this in the core meeting and we agreed that it would be much better if we have the fix for #2417 and #3098 in a different commit, just for documentation in case someone cannot update and wants to just understand what changed to backport that in their own application. Do you think it's doable?

Yep, makes sense! I'm on it

cedum commented on core/app/models/spree/address.rb:
```diff
@@ -253,9 +276,17 @@ def state_validate
+          caller
+      end
 
+      return unless country
```
ok boss

cedum commented:
I've extracted the fix for #3098 into a separate PR (merged into master), this way it should be easier to backport the fix on stores unable to upgrade. 
I've just rebased also this PR and I think it's ready for another review 🙏

kennyadsl reviewed (approved):
👍 Thanks @cedum !

aldesantis commented on core/lib/spree/core/engine.rb:
```diff
@@ -74,6 +74,15 @@ class Engine < ::Rails::Engine
+
+        if Spree::Config.use_legacy_address_state_validator != false
+          Spree::Deprecation.warn(
+            'Spree::Config.use_legacy_address_state_validator set to true is ' \
+            'deprecated and will not be possibile in Solidus 3.0. Please switch this ' \
+            'value to false and check that everything works as expected.',
```
Can we improve this message to explain that validation will be done through an external class moving forward?

cedum replied:
@aldesantis what do you think about this message?

> DEPRECATION WARNING: Spree::Config.use_legacy_address_state_validator set to true has been deprecated and will be removed in Solidus 3.0. The Spree::Address state validation has been extracted into a configurable external class. Switch Spree::Config.use_legacy_address_state_validator to true to start using the external validator class.

aldesantis replied:
@cedum sounds great!

aldesantis reviewed (changes requested):
@cedum just left a comment on a deprecation message, but other than that it's good to go!

spaghetticode reviewed (approved):
@cedum thank you 👍
